rag-openai-qdrant/rag-openai-qdrant.ipynb

```diff
@@ -0,0 +1,672 @@
+    "    completion = openai.ChatCompletion.create(\n",
+    "        model=\"gpt-3.5-turbo\",\n",
+    "        messages=[\n",
+    "            {\"role\": \"user\", \"content\": metaprompt},\n",
+    "        ],\n",
+    "        timeout=10.0,\n",
+    "    )\n",
+    "    return completion[\"choices\"][0][\"message\"][\"content\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "86fecd76-9a0b-4ad1-9097-b1d292a618ac",
+   "metadata": {},
+   "source": [
+    "We can ask our questions more easily."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "id": "aa0fdead-a115-4fcd-88dc-5cc718dc0544",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\"The stack for a web API can vary depending on the specific requirements and technologies used, but a common stack for a web API could include the following components:\\n\\n1. FastAPI: As mentioned in the context, FastAPI is a modern web framework for building APIs with Python.\\n\\n2. NGINX: NGINX can be used as a reverse proxy server to handle incoming requests and distribute them to the appropriate FastAPI instances. NGINX is known for its high performance and can help improve the scalability and reliability of the API.\\n\\n3. Docker: Docker can be utilized to containerize the FastAPI application along with any necessary dependencies. This allows for easy deployment and scalability, as Docker containers can be easily replicated and managed.\\n\\nIn addition to these components, the stack may also include other elements such as a database (e.g. PostgreSQL, MySQL) for storing and retrieving data, a cache (e.g. Redis, Memcached) for improving performance, and a message broker (e.g. RabbitMQ, Apache Kafka) for handling asynchronous communication.\\n\\nIt's important to note that the specific stack will depend on the requirements and constraints of the project, and there may be various alternative technologies and configurations that can be used.\""
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "rag(\"What can the stack for a web api look like?\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "7324c127-c140-410a-ab19-87a5babce023",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\"I don't know.\""
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "rag(\"Where is the nearest grocery store?\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6fe56730-ed41-42c1-9c33-de3849c60b65",
+   "metadata": {},
+   "source": [
+    "Our model can now utilize private knowledge and even admit it doesn't know the answer if it cannot be given based on the provided context. This is a pretty useful mechanism to mitigate the risks of hallucinations in Large Language Models."
```

Comment:
```suggestion
    "Our model can now:

1. Take advantage of the knowledge in our vector datastore 
2. Answer, based on the provided context, that it can not provide an answer. 

We have just shown a  useful mechanism to mitigate the risks of hallucinations in Large Language Models."

```